av1an-core/src/grain.rs

```diff
@@ -42,19 +85,27 @@ impl TransferFunction {
   #[allow(clippy::wrong_self_convention)]
   pub fn from_linear(self, x: f32) -> f32 {
     match self {
-      TransferFunction::BT1886 => x.powf(1. / 2.8),
+      TransferFunction::BT1886 => {
+        // Scale to a raw cd/m^2 value
+        let luma = x * BT1886_WHITEPOINT;
+
+        // The inverse of the `to_linear` formula:
+        // `(L / α)^(1 / λ) - β = x`
+        (luma / bt1886_alpha()).powf(1.0 / BT1886_GAMMA) - bt1886_beta()
+      }
       TransferFunction::SMPTE2084 => {
+        if x == 0.0 {
```

Comment:
Floating point comparisons to 0 should probably be done against an epsilon, like `x.abs() < f32::EPSILON`